### ISSUE TITLE
Headers with underscores should not be allowed in mock requests (since they are not allowed in real HTTP requests)

### ISSUE THREAD

**Issue by pawel-baster** [laravel/framework#46289]
- Laravel Version: 9.9.0
- PHP Version: 8.1.9
- Database Driver & Version: not relevant

### Description:

Headers with underscores in the field names (e.g. `api_key`) are generally filtered out when using Request::header(). However, they are **not** filtered out if a request is made from a unit test, what allows bugs to go unnoticed. 

### Steps To Reproduce:

In the following example a unit test is supposed to make sure that it is possible to authenticate a request with an authentication header. And as long as the unit test uses `$this->get()` to test the request, no error will be reported. This is unexpected, because in fact it is impossible to authenticate even when the header is valid, as proven by the subsequent failing `Http::get()` request.

Given an action defined as:
```
class UserController extends Controller {
    public function userAction(Request $request) {
        Log::info('header: ' . $request->header('api_key'));
        if ($request->header('api_key') !== 'valid-key') {
            return response('invalid key', 403);
        }

        return response('ok', 200);
    }
}
```
accessible in the unit test as `$url`, the following test does not pass:
```
class UserControllerTest extends TestCase
    public function testHeaders()
    {
        $headers = ['api_key' => 'valid-key'];
        $response = $this->withHeaders($headers)->get($url);
        $response->assertStatus(200); 
        // ^ this passes, because in UserController $request->header('api_key') === 'valid-key'

        $response = Http::withHeaders($headers)->get($url);
        $this->assertEquals(200, $response->status()); 
        // ^ this fails, because in UserController $request->header('api_key') === null, so the status code is 403
    }
}
```
Logs:
```
==> storage/logs/test.log <==
[2023-02-27 21:45:06] testing.INFO: header: valid-key  

==> storage/logs/laravel.log <==
[2023-02-27 21:45:06] local.INFO: header:   
```

**Comment by driesvints** [comment]
I don't think they're filtered out? Underscores in headers are allowed which is why tests allow them to pass. Most likely there's a setting in your http server that doesn't allow them.

See for example: https://stackoverflow.com/questions/22856136/why-do-http-servers-forbid-underscores-in-http-header-names

**Comment by pawel-baster** [comment]
@driesvints Ok, so it seems that headers with underscores are theoretically allowed, but in practice are not available in laravel while using default settings of both apache and nginx. Nginx drops them and on apache they are available when using `apache_request_headers()` or `get_all_headers()`, but not when using `$request->header()`. Apparently they are lost while the `$_SERVER` array is populated. I've just made a test on apache and these are the results of these different functions for a single request containing headers `api_key` and `api_secret`:
```
headers in apache_request_headers(): 
> ["Host","Accept","api_key","api_secret","tracestate","newrelic","traceparent"] 
headers in getallheaders():
> ["Host","Accept","api_key","api_secret","tracestate","newrelic","traceparent"] 
headers in $request->header():
> ["host","accept","tracestate","newrelic","traceparent"] 
headers in $_SERVER[HTTP_*]:
> ["HTTP_HOST","HTTP_ACCEPT","HTTP_TRACESTATE","HTTP_NEWRELIC","HTTP_TRACEPARENT"]
```
I've noticed the problem after replacing `apache_request_headers()` with `$request->header()`. My unit tests still passed, but the feature was broken for actual users.
